File: bridgeAddon/addon/globalPlugins/nvdaMcpBridge/adapters/json_lines_channel.py
# nvdaMcpBridge adapters -- JsonLinesChannel: the MessageChannel over a Transport.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
#
# ROLE: adapter. IMPLEMENTS the MessageChannel domain port.
# DEPENDS ON: the Transport adapter seam (adapters/ports/transport.py) -- never
#             on a concrete transport, which is what keeps this testable against
#             scripted bytes while the socket underneath stays a dumb leaf.
# BUILT BY: wiring.serve_transport (which pairs it with a concrete Transport).
#
# Every wire concern the domain must not know about lives here: reassembling
# chunks into newline-delimited frames, and JSON encode/decode via the shared
# protocol module. No NVDA import, so this stays fully type-checked.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .. import protocol
from ..domain.ports.message_channel import TIMEOUT, ChannelClosed, MessageChannel, Timeout

if TYPE_CHECKING:
	from .ports.transport import Transport


class _LineReader:
	"""Private to this adapter: reassembles chunks into complete lines."""

	def __init__(self) -> None:
		self._buffer = bytearray()

	def feed(self, chunk: bytes) -> None:
		self._buffer.extend(chunk)

	def next_line(self) -> bytes | None:
		"""Pop one complete line (without the newline), or ``None`` if incomplete."""
		newline = self._buffer.find(b"\n")
		if newline < 0:
			return None
		line = bytes(self._buffer[:newline])
		del self._buffer[: newline + 1]
		return line


class JsonLinesChannel(MessageChannel):
	"""Newline-delimited JSON messages over a byte transport."""

	def __init__(self, transport: Transport) -> None:
		self._transport = transport
		self._reader = _LineReader()

	def read_message(self) -> dict[str, Any] | Timeout:
		"""Next decoded message, or :data:`TIMEOUT`.

		Drains any line already buffered before touching the transport, so a
		message that already arrived is never lost to a timeout. Raises
		:class:`ChannelClosed` at EOF and
		:class:`protocol.ValidationError` for an unreadable line.
		"""
		while True:
			line = self._reader.next_line()
			if line is not None:
				return protocol.decode_message(line)
			try:
				chunk = self._transport.recv()
			except TimeoutError:
				return TIMEOUT
			if chunk == b"":
				raise ChannelClosed
			self._reader.feed(chunk)

	def write(self, message: Any) -> None:
		self._transport.sendall(protocol.encode_message(message))

	def close(self) -> None:
		self._transport.close()

File: bridges/voiceover/Sources/VoiceOverBridgeAdapters/JsonLinesChannel.swift
// ROLE: adapter implementing the MessageChannel port as newline-delimited JSON over a byte transport.
// BUILT BY: Wiring, which pairs it with whatever transport the accept produced.
// Drain a buffered line before reading the transport: two frames can arrive in one read, and the second would otherwise wait out an idle timeout.

import Foundation
import ScreenReaderWire
import VoiceOverBridgeDomain

public final class JsonLinesChannel: MessageChannel {
	private let transport: any Transport
	private let reader = LineReader()

	public init(transport: any Transport) {
		self.transport = transport
	}

	public func read() throws -> ChannelRead {
		while true {
			if let line = reader.nextLine() {
				return .message(try decode(line))
			}
			let chunk: Data
			do {
				chunk = try transport.receive()
			} catch is PollTimeout {
				return .timedOut
			}
			if chunk.isEmpty {
				throw ChannelClosed()
			}
			reader.feed(chunk)
		}
	}

	public func write(_ response: Response) throws {
		var data = try JSONEncoder().encode(response)
		data.append(0x0A)
		try transport.sendAll(data)
	}

	public func close() {
		transport.close()
	}

	/// A line that is not a JSON object has no `id` to answer, so it is rejected here as a protocol fault.
	private func decode(_ line: Data) throws -> [String: JSONValue] {
		let value: JSONValue
		do {
			value = try JSONDecoder().decode(JSONValue.self, from: line)
		} catch {
			throw ValidationError(path: "", reason: "line is not JSON: \(error)")
		}
		guard case .object(let fields) = value else {
			throw ValidationError(path: "", reason: "line is not a JSON object")
		}
		return fields
	}
}

final class LineReader {
	private var buffer = Data()

	func feed(_ chunk: Data) {
		buffer.append(chunk)
	}

	/// Pop one complete line, without its newline, or nil when none is complete.
	func nextLine() -> Data? {
		guard let newline = buffer.firstIndex(of: 0x0A) else { return nil }
		let line = buffer[buffer.startIndex..<newline]
		buffer = buffer[buffer.index(after: newline)...]
		return Data(line)
	}
}
